src/main/java/com/amazon/opendistroforelasticsearch/ad/task/ADTaskManager.java

```diff
@@ -260,33 +295,52 @@ protected void forwardToCoordinatingNode(
      * For historical detector, will set its AD task as cancelled.
      *
      * @param detectorId detector id
+     * @param historical stop historical analysis or not
      * @param handler AD job action handler
      * @param user user
      * @param transportService transport service
      * @param listener action listener
      */
     public void stopDetector(
         String detectorId,
+        boolean historical,
         IndexAnomalyDetectorJobActionHandler handler,
         User user,
         TransportService transportService,
         ActionListener<AnomalyDetectorJobResponse> listener
     ) {
-        getDetector(
-            detectorId,
-            // stop realtime detector job
-            (detector) -> handler.stopAnomalyDetectorJob(detectorId),
-            // stop historical detector AD task
-            (detector) -> getLatestADTask(
-                detectorId,
-                (task) -> stopHistoricalDetector(detectorId, task, user, listener),
-                transportService,
-                listener
-            ),
-            listener
-        );
+        getDetector(detectorId, (detector) -> {
+            if (historical) {
+                // stop historical detector AD task
+                getLatestADTask(detectorId, (task) -> stopHistoricalAnalysis(detectorId, task, user, listener), transportService, listener);
+            } else {
+                // stop realtime detector job
+                handler.stopAnomalyDetectorJob(detectorId);
+            }
+        }, listener);
+    }
+
+    public <T> void getDetector(String detectorId, Consumer<AnomalyDetector> consumer, ActionListener<T> listener) {
+        GetRequest getRequest = new GetRequest(AnomalyDetector.ANOMALY_DETECTORS_INDEX).id(detectorId);
+        client.get(getRequest, ActionListener.wrap(response -> {
+            if (!response.isExists()) {
+                listener.onFailure(new OpenSearchStatusException("AnomalyDetector is not found", RestStatus.NOT_FOUND));
+                return;
+            }
+            try (XContentParser parser = createXContentParserFromRegistry(xContentRegistry, response.getSourceAsBytesRef())) {
+                ensureExpectedToken(XContentParser.Token.START_OBJECT, parser.nextToken(), parser);
+                AnomalyDetector detector = AnomalyDetector.parse(parser, response.getId(), response.getVersion());
+
+                consumer.accept(detector);
```

Comment:
I didn't find you have listener.onResponse after this.

Reply:
`listener.onResponse` will be in `consumer`.